concurrent/actor/eventloop/EventLoop.java

```diff
@@ -45,52 +49,60 @@ public EventLoop(ThreadFactory factory) {
         thread.start();
     }
 
-    public void submit(Runnable job) {
-        eventQueue.offer(job);
+    public void submit(Runnable job, Consumer<Exception> onError) {
+        events.offer(Collections.pair(job, onError));
     }
 
-    public EventLoop.ScheduledJob submit(long millis, Runnable job) {
-        return new ScheduledJob(millis, job);
+    public EventLoop.ScheduledJob submit(long millis, Runnable job, Consumer<Exception> errorHandler) {
+        return new ScheduledJob(millis, job, errorHandler);
+    }
+
+    public void await() throws InterruptedException {
+        thread.join();
     }
 
     public void stop() throws InterruptedException {
-        submit(() -> state = State.STOPPED);
+        submit(() -> state = State.STOPPED, errorHandler);
         thread.join();
     }
 
     private void loop() {
-        LOG.debug("Starting EventLoop");
+        LOG.debug("Started");
         state = State.RUNNING;
         try {
             while (state == State.RUNNING) {
                 // TODO review performance, might want to batch some events from the regular queue before checking timers
                 long currentTime = GlobalSystem.time();
-                Runnable runnable = timerQueue.poll(currentTime);
+                Runnable runnable = timers.poll(currentTime);
                 if (runnable != null) {
                     runnable.run();
                 } else {
-                    Runnable event = eventQueue.poll(timerQueue.timeToNext(currentTime), TimeUnit.MILLISECONDS);
+                    Pair<Runnable, Consumer<Exception>> event = events.poll(timers.timeToNext(currentTime), TimeUnit.MILLISECONDS);
                     if (event != null) {
-                        event.run();
+                        try {
+                            event.first().run();
+                        } catch (Exception e) {
+                            event.second().accept(e);
+                        }
                     }
                 }
             }
-        } catch (Exception ex) {
-            LOG.error("EventLoop error", ex);
+        } catch (Exception e) {
+            errorHandler.accept(e);
```

Comment:
this looks odd - all error handling is now happening inside the `while` loop, so we shouldn't have to have any sophisticated error handling outside the loop

Reply:
agreed - does code outside the `while` loop can create error?

Reply:
Fixed